# Pair Programming 
## Lección 11 - EDA II - Nulos y Valores extremos
En esta lección hemos aprendido qué son los valores nulos y los outliers o valores faltantes. Además hemos aprendido una librería super molona que nos permite hacer una exploración de nuestro dataset muy profunda.

### Hipótesis
Antes de seguir, recordamos las preguntas que nos planteamos al principio del pair programming de EDA para dirigir nuestro análisis.

- ¿Es Australia el sitio más peligroso y letal para estar relajada en la playa?
- ¿Hay diferencias entre los paises en los ataques a hombres y mujeres?
- ¿Cuáles son las edades que más sufren ataques?
- Independientemente de la edad, ¿sufren los hombres más ataques que las mujeres?
- ¿En qué mes ocurren más ataques?
- ¿Cuál es la relación entre la especie y el tipo de ataque (si es fatal o no)?
- ¿Cómo han evolucionado los ataques a lo largo del tiempo?

De todo esto, nos damos cuenta que solo tenemos que limpiar algunas de las columnas, en concreto age, species, country, fatal, year, sex. Reducimos esto a una tabla para saber qué tenemos ya limpito y qué no. Actualizamos esta tabla ya que en el ejercicio de pair de Limpieza I ya dejamos algunas columnas limpitas:

variable        ¿Está limpia?

age             ❌ esta en formato string cuando debería ser integer, y en algunos casos tenemos rangos de edad

species         ❌ es un jaleo! Debemos unificar los nombres y reducir a las especies más importantes

country         ❌ los paises están en mayúsculas, algunos se repiten con algunos cambios

fatal           ✔️ la limpiamos en el pair de Pandas V

year            ❌ es una columna de tipo float, deberíamos convertirla a integer

sex             ✔️ la limpiamos en el pair de Pandas V

fecha           ✔️ la limpiamos en el pair de Pandas V

In [1]:
import pandas as pd
import sidetable

In [2]:
df = pd.read_csv("limpieza1.csv", index_col=0)

In [3]:
df.head(2)

year        type country  activity age     species_  mes fatal sex
0  2018     Boating     usa  Paddling  57  White shark  Jun     N   F
1  2018  Unprovoked  brazil  Swimming  18  Tiger shark  Jun     Y   M

### Objetivos
Usad la librería sidetable para:

1. Explorar los valores nulos.

In [4]:
df.stb.missing()

missing  total    percent
mes           214   1667  12.837433
age           150   1667   8.998200
species_      123   1667   7.378524
fatal          97   1667   5.818836
activity       31   1667   1.859628
sex            15   1667   0.899820
country        10   1667   0.599880
year            0   1667   0.000000
type            0   1667   0.000000

Explorando los nulos de nuestro dataframe vemos que:

.- mes = Nos encontramos un 12.83% de nulos.

.- age = Nos encontramos un 8.99% de nulos.

.- species_ = Nos encontraoms un 7.3% de nulos.

.- fatal = Nos encontramos un 5.8% de nulos.

.- activity = Nos encontramos un 1.8% de nulos.

.- sex, country = No llegan al 1% de nulos.

.- year = No tiene nulos.

En conclusión, visualizamos que no hay un porcentaje elevado de nulos en nuestro dataframe.

2. Explorad la columna de country y reportad los resultados solo para aquellas categorías que incluyan el 70% de los datos.

In [5]:
df.stb.freq(["country"])

country  count    percent  cumulative_count  \
0                        usa    740  44.659022               740   
1                  australia    340  20.519010              1080   
2               south africa    234  14.121907              1314   
3                    bahamas     29   1.750151              1343   
4                new zealand     21   1.267351              1364   
..                       ...    ...        ...               ...   
90  between portugal & india      1   0.060350              1653   
91                    belize      1   0.060350              1654   
92                   bahrein      1   0.060350              1655   
93                     asia?      1   0.060350              1656   
94                 argentina      1   0.060350              1657   

    cumulative_percent  
0            44.659022  
1            65.178033  
2            79.299940  
3            81.050091  
4            82.317441  
..                 ...  
90           99.758600  
91           99.818950  
92           99.879300  
93           99.939650  
94          100.000000  

[95 rows x 5 columns]

En esta visualización vemos que el 70% de nuestros datos en la columna "country" se agrupan en solo dos países: usa y australia.

In [6]:
df.stb.freq(["country"], thresh= 70, other_label="others countrys")

country  count    percent  cumulative_count  cumulative_percent
0              usa    740  44.659022               740           44.659022
1        australia    340  20.519010              1080           65.178033
2  others countrys    577  34.821967              1657          100.000000

Con el parámetro thresh podemos visualizar mejor los dos países que se encuentran en el 70% de nuestros datos. Los demás países los agrupamos en la etiqueta "others countrys"

3. Explorad la columna country en relación al sexo, filtrando de nuevo por aquellas categorías que incluyan el 70% de los datos.

In [7]:
df.stb.freq(["country", "sex"], thresh= 70, other_label="others countrys")

country              sex  count    percent  cumulative_count  \
0              usa                M    608  36.982968               608   
1        australia                M    306  18.613139               914   
2     south africa                M    224  13.625304              1138   
3  others countrys  others countrys    506  30.778589              1644   

   cumulative_percent  
0           36.982968  
1           55.596107  
2           69.221411  
3          100.000000

Visualizando nuestro sidetable vemos que solo hay tres países que se agrupen en el 70% de los datos (usa, australia y south africa) y solo tienen registro de ataques a hombres.

4. Haced los mismo para la columna activity en relación a la columna fatal. En este caso filtrad y mostrad los datos que incluyan el 60% de los datos.

In [8]:
df.stb.freq(["activity", "fatal"], thresh= 60, other_label="others activities")

activity              fatal  count    percent  cumulative_count  \
0             Surfing                  N    379  24.514877               379   
1            Swimming                  N    139   8.990944               518   
2        Spearfishing                  N     85   5.498060               603   
3            Swimming                  Y     78   5.045278               681   
4             Fishing                  N     50   3.234153               731   
5              Wading                  N     43   2.781371               774   
6            Standing                  N     30   1.940492               804   
7          Snorkeling                  N     29   1.875809               833   
8             Surfing                  Y     26   1.681759               859   
9       Body boarding                  N     24   1.552393               883   
10             Diving                  N     19   1.228978               902   
11       Scuba diving                  N     17   1.099612               919   
12  others activities  others activities    627  40.556274              1546   

    cumulative_percent  
0            24.514877  
1            33.505821  
2            39.003881  
3            44.049159  
4            47.283312  
5            50.064683  
6            52.005175  
7            53.880983  
8            55.562743  
9            57.115136  
10           58.344114  
11           59.443726  
12          100.000000

Visalizando nuestro dataframe con sidetable comprobamos que nos aparecen muchas actividades pero en relación a la columna "Fatal", las actividades que tienen accidentes mortales son "swimming" y "surfing", mientras que las otras sufres ataques de tiburones pero con poca probabilidad de mortalidad.

En el caso de ataques vemos que el surfing tiene un 24.51% de ataques no mortales de tiburones frente a un 1.68% de ataques mortales de tiburones. En cuanto al swimming vemos que tiene un 8.99% de ataques no mortales de tiburones frente a un 5.04% de ataques mortales de tiburones. 

Sacamos en conclusión que hay pocos ataques mortales realizando las diferentes actividades.

5. Por último, realizad este análisis para la columna fecha (la que creamos en el último ejercicio de pair programming de Pandas) en relación a la columna fatal. Mostrad solo los resultados para las categorías que incluyan el 60% de los datos.

In [9]:
df.stb.freq(["mes", "fatal"], thresh= 60)

mes   fatal  count    percent  cumulative_count  cumulative_percent
0     Oct       N    128   9.302326               128            9.302326
1     Aug       N    121   8.793605               249           18.095930
2     Sep       N    115   8.357558               364           26.453488
3     Jun       N    106   7.703488               470           34.156977
4     Jul       N    106   7.703488               576           41.860465
5     Apr       N     93   6.758721               669           48.619186
6     May       N     91   6.613372               760           55.232558
7  others  others    616  44.767442              1376          100.000000

Visualizando este sidetable en relación a la columna "mes" y "fatal" vemos que en los meses más cálidos comprendidos entre Abril y Octubre es cuando se dan más ataques de tiburones.

📌 Nota Igual que siempre, no nos vale solo con picar el código, queremos que nos incluyais las conclusiones que extraeis.

### Conclusiones
Si bien es cierto que la lección de hoy ha sido muy teórica, en esta parte los ejercicios no serán tanto de programar sino de pensar 🤔. ¿Qué tendréis que hacer?

- En relación a los valores nulos:
    - Identificar las columnas que tienen nulos (lo hicimos en el ejercicio de pair programming EDA I y en el de hoy con el sidetable).
    - Debatid entre vosotras qué haríais con los valores nulos, ¿los quitaríais o los reemplazaríais con un valor que nosotras decidiremos?. Cread una celda de markdown en vuestro jupyter y escribid vuestras conclusiones.

In [10]:
df.stb.missing()

missing  total    percent
mes           214   1667  12.837433
age           150   1667   8.998200
species_      123   1667   7.378524
fatal          97   1667   5.818836
activity       31   1667   1.859628
sex            15   1667   0.899820
country        10   1667   0.599880
year            0   1667   0.000000
type            0   1667   0.000000

Debatiendo entre nosotras sobre los valores nulos vemos que la columna que presenta un porcentaje mayor de nulos es "mes" y no supera el 13%, por lo tanto, pensamos que sería asumible eliminar los registros nulos y nos seguiríamos quedando con una parte bastante representativa de dichos datos para su análisis.

- En relación a los outliers, ya hemos ido conociendo un poco nuestro dataframe y de momento solo tenemos una columna de tipo numérico, por lo tanto, poco podemos hacer. Un poco más adelante, cuando hayamos hecho una serie de cambios sobre nuestras columnas podremos hacer una evaluación más profunda de estos.

Happy coding 🦈